### PR TITLE
Store persisted favicons outside the OS-clearable cache directory

### PR DESCRIPTION
**Pull request by theabhishekchandra**
Task/Issue URL: https://github.com/duckduckgo/Android/issues/6374
Tech Design URL (if applicable): 
API Proposals URL(s) (if applicable): None

### Description

Favicons for bookmarked and favorited sites were stored under `context.cacheDir` via `FileBasedFaviconPersister`. Android is permitted to automatically clear an app's cache directory under low storage conditions, so saved favicons would randomly disappear and fall back to a letter-avatar placeholder, even though the bookmark/favorite itself was untouched.

Per-tab temporary favicons (`FAVICON_TEMP_DIR`) are fine to lose and still use `cacheDir`. Persisted favicons for bookmarks/favorites (`FAVICON_PERSISTED_DIR`) and widget placeholders (`FAVICON_WIDGET_PLACEHOLDERS_DIR`) now use `context.filesDir`, which the OS does not automatically clear.

Favicons that existing installs already have under `cacheDir` are moved to `filesDir` the first time the persisted directory is accessed after the upgrade, so bookmark and favorite icons carry over without waiting for each site to be visited again. A file already written to the new location wins, leftover `.tmp` files from interrupted writes are skipped, and the legacy directory is deleted once moved, so later lookups only pay for a single `exists()` check. Re-fetching from the network on a miss only happens when the opt-in favicon fetching setting is enabled, so the migration is what keeps icons visible for most users.

### Steps to test this PR

_Existing favicons survive the upgrade_
- [x] On a build from `develop`, bookmark a site with a distinct favicon and confirm the file exists under `cache/favicons/`
- [x] Install this build over it and open Bookmarks
- [x] Confirm the favicon still displays, the file now lives under `files/favicons/`, and `cache/favicons/` is gone

_Favicon survives cache clearing_
- [x] Bookmark or favorite a site with a distinct favicon and confirm it displays correctly
- [x] Force-stop the app, then clear the app's cache via Android Settings > Apps > DuckDuckGo > Storage > Clear cache (or simulate OS-level cache eviction)
- [x] Relaunch the app and open Bookmarks/Favorites
- [x] Confirm the favicon still displays correctly instead of falling back to a letter avatar

### UI changes
No new UI. Screenshots below show the bug and the fix on a physical device (Vivo V2036, Android 13). Same steps on both builds: bookmark github.com, force-stop the app, delete the app's cache directory (what the OS does under storage pressure), relaunch.

| Before (develop) | After (this PR) |
| ------ | ----- |
| Bookmarks list falls back to the letter placeholder | Real favicon survives the cache clear |

Bookmarks before/after

Search and Favorites widget, same steps (widget loads the favicon through the updated FileProvider paths):

Widget before/after

---

> [!NOTE]
> **Medium Risk**
> Upgrade-time file migration and FileProvider path changes affect bookmark/favorite and widget icon loading; behavior is localized but storage failures could leave legacy cache until retry.
> 
> **Overview**
> **Persisted bookmark/favorite and widget placeholder favicons** no longer live under the OS-clearable `cacheDir`; only per-tab temp favicons stay in cache. `FileBasedFaviconPersister` routes `FAVICON_PERSISTED_DIR` and `FAVICON_WIDGET_PLACEHOLDERS_DIR` through `filesDir` and **migrates** any matching legacy tree from `cacheDir` on first access (synced, skips `.tmp`, keeps newer files in the new location, retries if a move fails).
> 
> **FileProvider** `provider_paths.xml` switches those favicon roots from `cache-path` to `files-path` so widgets and other URI consumers still resolve icons after the move.
> 
> Instrumented tests cover base-dir choice, migration edge cases (concurrency, partial failure, nested files), and updated test helpers to mirror the new layout.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5e135c6779282fbe1219520041775002d45716a2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by theabhishekchandra** [comment]
Hi @landomen, gentle ping on this one since you touched the widget favicon paths in #8315.

This fixes #6374, where bookmark and favorite favicons randomly revert to letter placeholders. Root cause is that persisted favicons live under `cacheDir`, which Android clears under storage pressure. The change moves persisted and widget favicons to `filesDir` and updates the matching FileProvider paths. Per-tab temp favicons stay in cache.

I've rebased onto current `develop`, reproduced the bug on a physical device by clearing the app cache, and confirmed the favicons survive after the fix, including the favorites widget. Before/after screenshots are in the description. Bugbot's earlier FileProvider finding is addressed in the second commit.

Could someone approve the workflows so CI runs, and let me know if you'd like anything changed? Happy to adjust.

File: app/src/main/java/com/duckduckgo/app/browser/favicon/FaviconPersister.kt [hunk 1]
```diff
@@ -281,7 +281,11 @@ class FileBasedFaviconPersister(
     }
 
     private fun faviconDirectory(directory: String): File {
-        return File(context.cacheDir, directory)
+        // FAVICON_TEMP_DIR holds per-tab favicons and is fine to live under cacheDir, which the OS
+        // may clear under storage pressure. FAVICON_PERSISTED_DIR (bookmarks/favorites) and
+        // FAVICON_WIDGET_PLACEHOLDERS_DIR must survive that, so they live in non-cache storage.
+        val baseDir = if (directory == FAVICON_TEMP_DIR) context.cacheDir else context.filesDir
+        return File(baseDir, directory)
```
[the remark this answers is not in the document]
Comment (theabhishekchandra) [reply app/src/main/java/com/duckduckgo/app/browser/favicon/FaviconPersister.kt]: Good catch, addressed in 4267e8a39c. Persisted and widget-placeholder favicons still under `cacheDir` are now moved to `filesDir` the first time the directory is resolved after the upgrade, and the legacy directory is removed once moved. Verified on a device by installing this branch over a `develop` build with a favicon in `cache/favicons`: the file moved with an identical checksum and the Bookmarks list and favorites widget showed it without a revisit. Instrumented tests cover the move, keeping a newer file already in the new location, nested folders, leftover `.tmp` files and concurrent access.

File: app/src/main/java/com/duckduckgo/app/browser/favicon/FaviconPersister.kt [hunk 2]
```diff
@@ -281,7 +282,47 @@ class FileBasedFaviconPersister(
...
+                        target.parentFile?.mkdirs()
+                        // A file already written by this version is newer than the legacy one, keep it.
+                        if (!target.exists() && !legacyFile.renameTo(target)) {
+                            legacyFile.copyTo(target)
+                        }
+                    }
+            }.onFailure {
+                logcat(LogPriority.WARN) { "FaviconPersister: failed to migrate legacy $directory favicons: ${it.message}" }
+            }
+            legacyDirectory.deleteRecursively()
```
[the remark this answers is not in the document]
Comment (theabhishekchandra) [reply app/src/main/java/com/duckduckgo/app/browser/favicon/FaviconPersister.kt]: Agreed, fixed in 5e135c6779. Each legacy file is now moved on its own and the legacy directory is only deleted once every file made it across; a failure part-way leaves the remaining files in place and the next lookup retries them. The copy fallback goes through a temp file that is removed on failure, so a truncated favicon is never left in the new location. Covered by an instrumented test that blocks one file's destination, checks the others migrated and the legacy directory survived, then unblocks it and checks the retry finishes the job. Also verified on a device with the real OS cache trim (`pm trim-caches`) and a 500-file legacy directory.
